lib/Exporter/Writer/KnpSnappyWriter.php

```diff
@@ -0,0 +1,221 @@
+</html>';
+
+    /**
+     * @throws \RuntimeException
+     *
+     * @param string $filename
+     * @param bool   $showHeaders
+     * @param bool   $showBorders
+     * @param array  $snappyOptions     KnpSnappy options (wkhtmltopdf)
+     * @param string $wkhtmltopdfBinary Path of the wkhtmltopdf binary
+     */
+    public function __construct(
+        $filename,
+        $wkhtmltopdfBinary,
+        $showHeaders = true,
+        $showBorders = true,
+        array $snappyOptions = array()
+    ) {
+        $this->filename = $filename;
+        $this->showHeaders = $showHeaders;
+        $this->showBorders = $showBorders;
+        $this->snappyOptions = $snappyOptions;
+        $this->wkhtmltopdfBinary = $wkhtmltopdfBinary;
+
+        if (is_file($filename)) {
+            throw new \RuntimeException(sprintf('The file "%s" already exists', $filename));
+        }
+
+        if (!is_executable($wkhtmltopdfBinary)) {
+            throw new \RuntimeException('The file "%s" doesn\'t exist or is not executable', $wkhtmltopdfBinary);
+        }
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function open()
+    {
+        $this->file = fopen($this->filename, 'w', false);
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function write(array $data)
+    {
+        $tmpFile = tmpfile();
+
+        if (true === $this->showBorders) {
+            $this->addBorders();
+        }
+
+        $this->addMarkupFromOptions();
+
+        if (true === $this->showHeaders) {
+            fwrite($tmpFile, $this->getHeaders($data));
+        }
+
+        fwrite($tmpFile, '<tr>');
+
+        foreach ($data as $value) {
+            fwrite($tmpFile, sprintf('<td>%s</td>', $value));
+        }
+
+        fwrite($tmpFile, '</tr>');
+
+        $tmpFileMetaData = stream_get_meta_data($tmpFile);
+        $tmpFilePath = $tmpFileMetaData['uri'];
+
+        fwrite($this->file, $this->generatePdf($tmpFilePath));
+
+        fclose($tmpFile);
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function close()
+    {
+        fclose($this->file);
+    }
+
+    /**
+     * Generates a PDF using KnpSnappy.
+     *
+     * @param string $htmlFilename.
+     *
+     * @return string The generated PDF content.
+     */
+    private function generatePdf($htmlFilename)
+    {
+        $snappy = new Pdf();
+        $snappy->setBinary($this->wkhtmltopdfBinary);
+
+        foreach ($this->snappyOptions as $key => $value) {
+            $snappy->setOption($key, $value);
+        }
+
+        $this->html = str_replace('[content]', file_get_contents($htmlFilename), $this->html);
+
+        return $snappy->getOutputFromHtml($this->html);
+    }
+
+    /**
+     * @param array $data
+     *
+     * @return string
+     */
+    private function getHeaders(array $data)
+    {
+        $row = '';
+
+        foreach ($data as $header => $value) {
+            $row .= sprintf('<th>%s</th>', $header);
+        }
+
+        return sprintf('<tr>%s</tr>', $row);
```

Comment:
> The code you given doesn't work as is.

That's why I said "something like this" ;)

It is in the natural order, and what you are doing is templating for a header, so you should use something that looks like a template. The code is simple at the moment, but if it becomes more complex, concatenation and sprintf will no longer be acceptable. That's why we don't use concatenation and sprintf for templates.